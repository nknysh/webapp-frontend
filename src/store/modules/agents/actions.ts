import { ICompany, ITravelAgent } from 'services/BackendApi';
export const GET_TRAVEL_AGENTS_REQUEST = 'agentsModule/GET_TRAVEL_AGENTS_REQUEST';
export const GET_TRAVEL_AGENTS_SUCCESS = 'agentsModule/GET_TRAVEL_AGENTS_SUCCESS';
export const GET_TRAVEL_AGENTS_FAILURE = 'agentsModule/GET_TRAVEL_AGENTS_FAILURE';
export const SELECTED_TA_CHANGE = 'agentsModule/SELECTED_TA_CHANGE';
export const SHOW_TA_DROPDOWN = 'agentsModule/SHOW_TA_DROPDOWN';
export const SEARCH_TA_BY_NAME_CHANGE = 'agentsModule/SEARCH_TA_BY_NAME_CHANGE';
export const GET_COMPANIES_REQUEST = 'agentsModule/GET_COMPANIES_REQUEST';
export const GET_COMPANIES_SUCCESS = 'agentsModule/GET_COMPANIES_SUCCESS';
export const GET_COMPANIES_FAILURE = 'agentsModule/GET_COMPANIES_FAILURE';
export const SELECTED_COMPANY_CHANGE = 'agentsModule/SELECTED_COMPANY_CHANGE';
export const SHOW_COMPANY_DROPDOWN = 'agentsModule/SHOW_COMPANY_DROPDOWN';
export const SEARCH_COMPANY_BY_NAME = 'agentsModule/SEARCH_COMPANY_BY_NAME';

export type GetTravelAgentRequestAction = ReturnType<typeof getTravelAgentsRequestAction>;
export const getTravelAgentsRequestAction = (forceLoad?: boolean) => ({
  type: GET_TRAVEL_AGENTS_REQUEST as typeof GET_TRAVEL_AGENTS_REQUEST,
  forceLoad,
});

export type GetTravelAgentSuccessAction = ReturnType<typeof getTravelAgentsSuccessAction>;
export const getTravelAgentsSuccessAction = (agents: ITravelAgent[]) => ({
  type: GET_TRAVEL_AGENTS_SUCCESS as typeof GET_TRAVEL_AGENTS_SUCCESS,
  agents,
});

export type GetTravelAgentFailureAction = ReturnType<typeof getTravelAgentsFailureAction>;
export const getTravelAgentsFailureAction = (error: string) => ({
  type: GET_TRAVEL_AGENTS_FAILURE as typeof GET_TRAVEL_AGENTS_FAILURE,
  error,
});

export type SelectedTaChangeAction = ReturnType<typeof selectedTaChangeAction>;
export const selectedTaChangeAction = (value: ITravelAgent | null) => ({
  type: SELECTED_TA_CHANGE as typeof SELECTED_TA_CHANGE,
  value,
});

export type ShowTaDropdownAction = ReturnType<typeof showTaDropdownAction>;
export const showTaDropdownAction = (value: boolean) => ({
  type: SHOW_TA_DROPDOWN as typeof SHOW_TA_DROPDOWN,
  value,
});

export type SearchTaByNameChangeAction = ReturnType<typeof searchTaByNameChangeAction>;
export const searchTaByNameChangeAction = (value: string) => ({
  type: SEARCH_TA_BY_NAME_CHANGE as typeof SEARCH_TA_BY_NAME_CHANGE,
  value,
});


export type GetCompaniesRequestAction = ReturnType<typeof getCompaniesRequestAction>;
export const getCompaniesRequestAction = () => ({
  type: GET_COMPANIES_REQUEST as typeof GET_COMPANIES_REQUEST,
});

export type GetCompaniesSuccessAction = ReturnType<typeof getCompaniesSuccessAction>;
export const getCompaniesSuccessAction = (companies: ICompany[]) => ({
  type: GET_COMPANIES_SUCCESS as typeof GET_COMPANIES_SUCCESS,
  companies,
});

export type GetCompaniesFailureAction = ReturnType<typeof getCompaniesFailureAction>;
export const getCompaniesFailureAction = (error: string) => ({
  type: GET_COMPANIES_FAILURE as typeof GET_COMPANIES_FAILURE,
  error,
});

export type SelectedCompanyChangeAction = ReturnType<typeof selectedCompanyChangeAction>;
export const selectedCompanyChangeAction = (value: ICompany) => ({
  type: SELECTED_COMPANY_CHANGE as typeof SELECTED_COMPANY_CHANGE,
  value,
});

export type ShowCompanyDropdownAction = ReturnType<typeof showCompanyDropdownAction>;
export const showCompanyDropdownAction = (value: boolean) => ({
  type: SHOW_COMPANY_DROPDOWN as typeof SHOW_COMPANY_DROPDOWN,
  value,
});

export type SearchCompanyByNameAction = ReturnType<typeof searchCompanyByNameAction>;
export const searchCompanyByNameAction = (value: string) => ({
  type: SEARCH_COMPANY_BY_NAME as typeof SEARCH_COMPANY_BY_NAME,
  value,
});

export type AgentsModuleAction =
  | GetTravelAgentRequestAction
  | GetTravelAgentSuccessAction
  | GetTravelAgentFailureAction
  | SelectedTaChangeAction
  | ShowTaDropdownAction
  | SearchTaByNameChangeAction
  | GetCompaniesRequestAction
  | GetCompaniesSuccessAction
  | GetCompaniesFailureAction
  | SelectedCompanyChangeAction
  | ShowCompanyDropdownAction
  | SearchCompanyByNameAction;
