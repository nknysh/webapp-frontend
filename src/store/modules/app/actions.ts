export const SET_LATEST_APP_VERSION = 'app/SET_LATEST_APP_VERSION';

export type SetLatestAppVersion = ReturnType<typeof setLatestAppVersion>;
export const setLatestAppVersion = (latestAppVersion: string, isAppDeprecated: boolean) => ({
  type: SET_LATEST_APP_VERSION as typeof SET_LATEST_APP_VERSION,
  latestAppVersion,
  isAppDeprecated
});


export type AppAction =
 | SetLatestAppVersion
