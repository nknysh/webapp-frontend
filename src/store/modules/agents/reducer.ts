import * as Actions from './actions';
import { initialState, IAgentsModuleDomain } from './model';
import { getTaFullName } from 'store/utils';

export const agentsDomainReducer = (
  state: IAgentsModuleDomain = initialState,
  action: Actions.AgentsModuleAction
): IAgentsModuleDomain => {
  switch (action.type) {
    case Actions.GET_TRAVEL_AGENTS_REQUEST:
      return {
        ...state,
        requestPending: true,
      };

    case Actions.GET_TRAVEL_AGENTS_SUCCESS:
      return {
        ...state,
        requestPending: false,
        agents: action.agents,
      };

    case Actions.GET_TRAVEL_AGENTS_FAILURE:
      return {
        ...state,
        requestPending: true,
        agents: undefined,
        error: action.error,
      };

    case Actions.SELECTED_TA_CHANGE:
      return {
        ...state,
        selectedTa: action.value,
        taNameSearch: getTaFullName(action.value),
      };

    case Actions.SHOW_TA_DROPDOWN:
      return {
        ...state,
        showTaDropdown: action.value,
      };

    case Actions.SEARCH_TA_BY_NAME_CHANGE:
      return {
        ...state,
        taNameSearch: action.value,
      };


    case Actions.GET_COMPANIES_REQUEST:
      return {
        ...state,
        isFetchingCompanies: true,
      };

    case Actions.GET_COMPANIES_SUCCESS:
      return {
        ...state,
        isFetchingCompanies: false,
        companies: action.companies,
      };

    case Actions.GET_COMPANIES_FAILURE:
      return {
        ...state,
        isFetchingCompanies: true,
        companies: null,
        error: action.error,
      };

    case Actions.SELECTED_COMPANY_CHANGE:
      return {
        ...state,
        selectedCompany: action.value,
        companyNameSearch: action.value!.name,
        selectedTa: null
      };

    case Actions.SHOW_COMPANY_DROPDOWN:
      return {
        ...state,
        showCompanyDropdown: action.value,
      };

    case Actions.SEARCH_COMPANY_BY_NAME:
      return {
        ...state,
        companyNameSearch: action.value,
      };

    default:
      return state;
  }
};

export default agentsDomainReducer;
