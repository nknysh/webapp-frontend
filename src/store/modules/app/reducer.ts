import { initialState, AppState } from './model';
import * as Actions from './actions';
import produce from 'immer';

export const appReducer = (
  state: AppState = initialState,
  action: Actions.AppAction
) => {
  switch (action.type) {
    case Actions.SET_LATEST_APP_VERSION:
      return updateLatestAppVersionReducer(state, action);

    default:
      return state;
  }
};

export const updateLatestAppVersionReducer = (
  state: AppState,
  action: Actions.SetLatestAppVersion
): AppState => {
  return produce(state, draftState => {
    draftState.latestAppVersion = action.latestAppVersion;
    return draftState;
  });
};
