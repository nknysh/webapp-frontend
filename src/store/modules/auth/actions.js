import { prop, propEq, omit, propOr, mergeDeepLeft, equals } from 'ramda';
import { FORBIDDEN } from 'http-status';

import client from 'api/auth';
import userClient from 'api/users';
import { UserStatusTypes, AuthTypes } from 'config/enums';

import { genericAction, successAction, errorAction, errorFromResponse, storeReset } from 'store/common';
import { enqueueNotification } from 'store/modules/ui/actions';
import { getCurrentUserType } from './selectors';

export const AUTH_REQUEST = 'AUTH_REQUEST';
export const AUTH_CHECK = 'AUTH_CHECK';
export const AUTH_OK = 'AUTH_OK';
export const AUTH_RESET = 'AUTH_RESET';
export const AUTH_SET_TOKEN = 'AUTH_SET_TOKEN';
export const AUTH_SIGN_UP = 'AUTH_SIGN_UP';
export const AUTH_LOG_OUT = 'AUTH_LOG_OUT';
export const AUTH_PASSWORD_RESET = 'AUTH_PASSWORD_RESET';
export const AUTH_SET_PASSWORD = 'AUTH_SET_PASSWORD';
export const AUTH_COUNTRY_SET = 'AUTH_COUNTRY_SET';
export const AUTH_PASSWORD_UPDATE = 'AUTH_PASSWORD_UPDATE';

// Localstorage constants for auth
export const AUTH_TOKEN = 'authToken';
export const AUTH_USER = 'authUser';
export const AUTH_COUNTRY_CODE = 'authCountryCode';

/**
 * Auth reset action
 *
 * @returns {object}
 */
const authReset = () => ({
  type: AUTH_RESET,
});

/**
 * Set token action
 *
 * @param {string} token
 * @returns {object}
 */
export const setToken = token => ({
  type: AUTH_SET_TOKEN,
  payload: { token },
});

/**
 * Auth request action
 *
 * @param {*} value
 * @returns {object}
 */
export const authRequest = value => ({
  type: AUTH_REQUEST,
  payload: value,
});

/**
 * Auth password reset action
 *
 * @param {*} value
 * @returns {object}
 */
export const authPasswordReset = value => ({
  type: AUTH_PASSWORD_RESET,
  payload: value,
});

/**
 * Auth set password reset action
 *
 * @param {*} value
 * @returns {object}
 */
export const authSetPasswordReset = value => ({
  type: AUTH_SET_PASSWORD,
  payload: value,
});

/**
 * Auth ok action
 *
 * @param {*} values
 * @returns {object}
 */
export const authOk = values => ({
  type: AUTH_OK,
  payload: values,
});

/**
 * Auth sign up action
 *
 * @param {*} values
 * @returns {object}
 */
export const authSignUp = values => ({
  type: AUTH_SIGN_UP,
  payload: values,
});

/**
 * Auth log out action
 *
 * @param {string} token
 * @returns {object}
 */
export const authLogOut = token => ({
  type: AUTH_LOG_OUT,
  payload: token,
});

/**
 * Auth set country action
 *
 * @param {string} countryCode
 * @returns {object}
 */
export const authSetCountryAction = countryCode => ({
  type: AUTH_COUNTRY_SET,
  payload: countryCode,
});

/**
 * Set remembered token
 *
 * @param {string} token
 */
export const setRememberedToken = token => localStorage.setItem(AUTH_TOKEN, token);

/**
 * Set remembered user
 *
 * @param {object} user
 */
export const setRememberedUser = user => localStorage.setItem(AUTH_USER, JSON.stringify(user));

/**
 * Set remembered country
 *
 * @param {string} countryCode
 */
export const setRememberedCountry = countryCode => localStorage.setItem(AUTH_COUNTRY_CODE, countryCode);

/**
 * Delete remembered token
 */
export const deleteRememberedToken = () => localStorage.removeItem(AUTH_TOKEN);

/**
 * Delete remembered user
 */
export const deleteRememberedUser = () => localStorage.removeItem(AUTH_USER);

/**
 * Delete remembered country
 */
export const deleteRememberedCountry = () => localStorage.removeItem(AUTH_COUNTRY_CODE);

/**
 * Auth set country action
 *
 * Sets the remembered country into local storage
 *
 * @param {string} countryCode
 * @returns {Function}
 */
export const authSetCountry = countryCode => dispatch => {
  setRememberedCountry(countryCode);
  dispatch(authSetCountryAction(countryCode));
};

/**
 * Get user action
 *
 * @param {Function} dispatch
 */
export const getUser = async dispatch => {
  dispatch(authRequest());

  try {
    const { data } = await client.getUser();
    dispatch(successAction(AUTH_REQUEST, data));
  } catch (e) {
    dispatch(errorFromResponse(AUTH_REQUEST, e, 'There was a problem fetching the current user.'));
  }
};

/**
 * Sign up action
 *
 * @param {object} values
 * @returns {Function}
 */
export const signUp = values => async dispatch => {
  dispatch(authSignUp(values));

  try {
    const { data } = await client.signUp({ data: { attributes: values } });
    dispatch(successAction(AUTH_SIGN_UP, data));
  } catch (e) {
    dispatch(errorFromResponse(AUTH_SIGN_UP, e));
  }
};

/**
 * Clear user action
 *
 * Removes user form local storage
 *
 * @param {Function} dispatch
 */
const clearUser = dispatch => {
  deleteRememberedToken();
  deleteRememberedUser();
  deleteRememberedCountry();

  dispatch(authReset());
  dispatch(successAction(AUTH_LOG_OUT, {}));
  dispatch(storeReset());
};

/**
 * Log out action
 *
 * Trrigger logout request
 *
 * @param {string} token
 * @returns {Function}
 */
export const logOut = token => async dispatch => {
  dispatch(authLogOut(token));

  clearUser(dispatch);

  try {
    await client.logOut();
  } catch (e) {
    return true;
  }
};

/**
 * Persist user action
 *
 * Sets user data to local storage
 *
 * @param {Function} dispatch
 * @param {object} data
 */
const persistUser = (dispatch, data) => {
  const userUuid = prop('uuid', data);

  // If a user is 'pending' then throw a forbidden error and return
  if (propEq('status', UserStatusTypes.PENDING, data)) {
    return dispatch(errorAction(AUTH_REQUEST, { status: FORBIDDEN, unverified: true }));
  }

  setRememberedToken(userUuid);
  setRememberedUser(data);
  setRememberedCountry(propOr('GB', 'countryCode', data));

  dispatch(setToken(userUuid));
};

/**
 * Log in action
 *
 * @param {object} values
 * @returns {Function}
 */
export const logIn = values => async dispatch => {
  dispatch(authRequest(omit(['password'], values)));

  try {
    const {
      data: { data },
    } = await client.logIn({ data: { attributes: values } });

    persistUser(dispatch, data);
    dispatch(successAction(AUTH_REQUEST, { user: { ...data } }));
  } catch (e) {
    dispatch(errorFromResponse(AUTH_REQUEST, e));
  }
};

/**
 * Reset password action
 *
 * @param {object} values
 * @returns {Function}
 */
export const resetPassword = values => async dispatch => {
  dispatch(authPasswordReset(values));

  try {
    await client.resetPassword({ data: { attributes: values } });
    dispatch(successAction(AUTH_PASSWORD_RESET, values));
  } catch (e) {
    dispatch(errorFromResponse(AUTH_SET_PASSWORD, e));
    throw e;
  }
};

/**
 * Set password action
 *
 * @param {object} values
 * @returns {Function}
 */
export const setPassword = values => async dispatch => {
  dispatch(authSetPasswordReset(omit(['values'], values)));

  try {
    await client.setPassword({ data: { attributes: values } });
    dispatch(successAction(AUTH_SET_PASSWORD, values));
  } catch (e) {
    dispatch(errorFromResponse(AUTH_SET_PASSWORD, e));
  }
};

/**
 * Auth check action
 *
 * @param {object} params
 * @returns {Function}
 */
export const authCheck = params => async (dispatch, getState) => {
  dispatch(genericAction(AUTH_CHECK));

  const role = getCurrentUserType(getState());

  // Associations for auth call are different per role
  // TAs : assignedSalesRepresentatives
  // Everyone else : assignedTravelAgents
  const authParams = mergeDeepLeft(params, {
    associations: equals(AuthTypes.TA, role) ? 'assignedSalesRepresentatives' : 'assignedTravelAgents',
  });

  try {
    const {
      data: { data },
    } = await userClient.me(authParams);
    dispatch(successAction(AUTH_CHECK, { user: { ...data } }));
    persistUser(dispatch, data);
  } catch (e) {
    clearUser(dispatch);

    throw e;
  }
};

/**
 * Update password action
 *
 * @param {object} body
 * @returns {Function}
 */
export const updatePassword = body => async dispatch => {
  dispatch(genericAction(AUTH_PASSWORD_UPDATE, body));

  try {
    await client.updatePassword({ data: { attributes: body } });
    dispatch(successAction(AUTH_SET_PASSWORD, body));
    dispatch(enqueueNotification({ message: 'Password has been updated.', options: { variant: 'success' } }));
  } catch (e) {
    dispatch(errorFromResponse(AUTH_PASSWORD_UPDATE, e));
  }
};
