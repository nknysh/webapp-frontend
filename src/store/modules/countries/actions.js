import { ap, values, path, prop, mergeDeepRight } from 'ramda';

import { index } from 'store/modules/indexes/actions';
import { successAction } from 'store/common/actions';

import schema from './schema';
import { getCountriesEntities } from './selectors';

export const SET_COUNTRIES = 'SET_COUNTRIES';

export const setCountriesAction = payload => ({
  type: SET_COUNTRIES,
  payload,
});

export const setCountries = data => (dispatch, getState) => {
  const prevData = getCountriesEntities(getState());
  const countries = mergeDeepRight(prevData, path(['entities', 'countries'], data));

  ap(
    [dispatch],
    [
      successAction(SET_COUNTRIES, data),
      index({
        index: 'countries',
        ref: prop('id', schema),
        fields: prop('index', schema),
        data: values(countries),
      }),
    ]
  );
};
