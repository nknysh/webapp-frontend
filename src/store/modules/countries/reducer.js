import { initialState, loadingReducer, successReducer, errorReducer } from 'store/common';
import { createReducer, getErrorActionName, getSuccessActionName } from 'store/utils';

import { SET_COUNTRIES } from './actions';

export default createReducer(
  {
    [SET_COUNTRIES]: loadingReducer,
    [getSuccessActionName(SET_COUNTRIES)]: successReducer,
    [getErrorActionName(SET_COUNTRIES)]: errorReducer,
  },
  initialState
);
