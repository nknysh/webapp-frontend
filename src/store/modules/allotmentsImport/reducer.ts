import { EImportEntity }  from '../importer/model';
import { importReducer } from '../importer/reducer';

export const allotmentsImport = importReducer(EImportEntity.ALLOTMENTS);
