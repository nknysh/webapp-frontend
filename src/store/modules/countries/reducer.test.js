import { setCountries } from './actions';
import reducer from './reducer';

describe('countries reducer', () => {
  it('should return the initial state', () => {
    expect(reducer(undefined, {})).toMatchSnapshot();
  });

  it('should handle SET_COUNTRIES', () => {
    const action = setCountries();
    const reducedState = reducer({}, action);

    expect(reducedState).toMatchSnapshot();
  });

  it('should handle SET_COUNTRIES_SUCCESS', () => {
    const action = setCountries();
    const reducedState = reducer({}, action);

    expect(reducedState).toMatchSnapshot();
  });

  it('should handle SET_COUNTRIES_ERROR', () => {
    const action = setCountries();
    const reducedState = reducer({}, action);

    expect(reducedState).toMatchSnapshot();
  });
});
