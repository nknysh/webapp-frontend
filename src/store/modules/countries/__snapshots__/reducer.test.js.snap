// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages reducer should handle SET_COUNTRIES 1`] = `Object {}`;

exports[`pages reducer should handle SET_COUNTRIES_ERROR 1`] = `Object {}`;

exports[`pages reducer should handle SET_COUNTRIES_SUCCESS 1`] = `Object {}`;
